pkg/upgrade/update.go

```diff
@@ -12,49 +12,83 @@ import (
 	"github.com/openshift/openshift-azure/pkg/api"
 	"github.com/openshift/openshift-azure/pkg/log"
 	"github.com/openshift/openshift-azure/pkg/util/azureclient"
+	"github.com/openshift/openshift-azure/pkg/util/managedcluster"
 )
 
 func (u *simpleUpgrader) Update(ctx context.Context, cs *api.OpenShiftManagedCluster, azuretemplate map[string]interface{}, deployFn api.DeployFn) error {
 	err := u.createClients(ctx, cs)
 	if err != nil {
-		return err
+		return &api.PluginError{Err: err, Step: api.PluginStepClientCreation}
 	}
-	// Deploy() may change the number of VMs.  If we can see that any VMs are
+	// deployFn() may change the number of VMs.  If we can see that any VMs are
 	// about to be deleted, drain them first.  Record which VMs are visible now
-	// so that we can detect newly created VMs and wait for them to become
-	// ready.
+	// so that we can detect newly created VMs and wait for them to become ready.
+	vmsBefore, err := u.getNodesAndDrain(ctx, cs)
+	if err != nil {
+		return &api.PluginError{Err: err, Step: api.PluginStepDrain}
+	}
+	err = deployFn(ctx, azuretemplate)
+	if err != nil {
+		return &api.PluginError{Err: err, Step: api.PluginStepDeploy}
+	}
+	err = u.InitializeCluster(ctx, cs)
+	if err != nil {
+		return &api.PluginError{Err: err, Step: api.PluginStepInitialize}
+	}
+	err = managedcluster.WaitForHealthz(ctx, cs.Config.AdminKubeconfig)
+	if err != nil {
+		return &api.PluginError{Err: err, Step: api.PluginStepWaitForWaitForOpenShiftAPI}
+	}
+	err = u.waitForNewNodes(ctx, cs, vmsBefore)
+	if err != nil {
+		return &api.PluginError{Err: err, Step: api.PluginStepWaitForNodes}
+	}
 
+	// For PP day 1, scale is permitted but not any other sort of update.  When
+	// we enable configuration changes and/or upgrades, uncomment this code.  At
+	// the same time, current thinking is that we will add a hash-based
+	// mechanism to avoid unnecessary VM rotations as well.
+	if os.Getenv("RUNNING_UNDER_TEST") != "" {
```

Comment:
@jim-minter @asalkeld as requested here: https://github.com/openshift/openshift-azure/issues/672#issuecomment-434546004 can we more granularity in the stage of error as well?